src/setup_payload/python/SetupPayload.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+#
+#    Copyright (c) 2022 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+import enum
+
+import Base38
+from bitarray import bitarray
+from stdnum.verhoeff import calc_check_digit
+
+# See section 5.1.4.1 Manual Pairing Code in the Matter specification v1.0
+MANUAL_DISCRIMINATOR_LEN = 4
+PINCODE_LEN = 27
+
+MANUAL_CHUNK1_DISCRIMINATOR_MSBITS_LEN = 2
+MANUAL_CHUNK1_DISCRIMINATOR_MSBITS_POS = 0
+MANUAL_CHUNK1_VID_PID_PRESENT_BIT_POS = MANUAL_CHUNK1_DISCRIMINATOR_MSBITS_POS + MANUAL_CHUNK1_DISCRIMINATOR_MSBITS_LEN
+MANUAL_CHUNK1_LEN = 1
+
+MANUAL_CHUNK2_DISCRIMINATOR_LSBITS_LEN = 2
+MANUAL_CHUNK2_PINCODE_LSBITS_LEN = 14
+MANUAL_CHUNK2_PINCODE_LSBITS_POS = 0
+MANUAL_CHUNK2_DISCRIMINATOR_LSBITS_POS = MANUAL_CHUNK2_PINCODE_LSBITS_POS + MANUAL_CHUNK2_PINCODE_LSBITS_LEN
+MANUAL_CHUNK2_LEN = 5
+
+MANUAL_CHUNK3_PINCODE_MSBITS_LEN = 13
+MANUAL_CHUNK3_PINCODE_MSBITS_POS = 0
+MANUAL_CHUNK3_LEN = 4
+
+MANUAL_VID_LEN = 5
+MANUAL_PID_LEN = 5
+
+# See section 5.1.3. QR Code in the Matter specification v1.0
+QRCODE_VERSION_LEN = 3
+QRCODE_DISCRIMINATOR_LEN = 12
+QRCODE_VID_LEN = 16
+QRCODE_PID_LEN = 16
+QRCODE_COMMISSIONING_FLOW_LEN = 2
+QRCODE_DISCOVERY_CAP_BITMASK_LEN = 8
+QRCODE_PADDING_LEN = 4
+QRCODE_VERSION = 0
+QRCODE_PADDING = 0
+
+
+class CommissioningFlow(enum.IntEnum):
+    Standard = 0,
+    UserIntent = 1,
+    Custom = 2
+
+
+class SetupPayload:
+    def __init__(self, discriminator, pincode, rendezvous=4, flow=CommissioningFlow.Standard, vid=0, pid=0):
+        self.long_discriminator = discriminator
+        self.short_discriminator = discriminator >> 8
+        self.pincode = pincode
+        self.rendezvous = rendezvous
+        self.flow = flow
+        self.vid = vid
+        self.pid = pid
+
+    @staticmethod
+    def parse_qrcode(qrcode):
+        bytes = Base38.decode(qrcode.replace('MT:', ''))
+        bytes.reverse()
+        qrcode_bit_string = ''.join(format(x, '08b') for x in bytes)
```

Comment:
could we use something like https://construct.readthedocs.io/en/latest/ for this parsing?

Having this cast and convert does not seem as user friendly.

Reply:
Specifically the qr code format could look something like:

```py
format = BitStruct(
    'version' / BitsInteger(3),
    'vendor_id' / BitsInteger(16),
    'product_id' / BitsInteger(16),
    'custom_flow' / BitsInteger(2),
    'discovery_capabilities' / BitsInteger(8),
    'discriminator' / BitsInteger(12),
    'passcode' / BitsInteger(27),
    'padding'  / Padding(4),
    'tlv' / GreedyRange(Byte),
)
```

Except the capabilties and custom flow and such could actually be switched to readable enums. This I can much easier validate against a spec instead of trying to figure out offsets in a binary string.

Reply:
Looks like a much cleaner solution. Thanks for pointing out.